Slow down and stop on red traffic lights in the waypoints

idanov opened the pull request:
Using the `rviz` I can confirm that the waypoints have their velocity setup correctly. However it seems that the DBW node or the waypoint follower are causing issues with runaway speed or awkward slowing down once the breaks are allowed.
Fixes #60

idanov committed:
Copy the lookahead more efficiently

idanov committed:
Limit the queue_size of all subscribers

idanov committed:
Visualise next TL in rviz

idanov committed:
Change the visualisation of waypoints in rviz

idanov committed:
Slow down and stop on red traffic lights

idanov committed:
Make sure that waypoint follower doesn't get flooded with messages

carlosgalvezp commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -29,43 +30,57 @@
 '''
 
 LOOKAHEAD_WPS = 50  # Number of waypoints we publish
+MAX_DECEL = 5.
```
Please get this from the Parameter Server, so that we don't have duplicates. Should be done in `__init__`, after `rospy.init_node`. For example:

        self.max_acc = rospy.get_param('/dbw_node/accel_limit')

carlosgalvezp commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -79,25 +94,53 @@ def publish(self):
 
             # This is needed for visualising in rviz
-            path = Path()
+            path = Marker()
```
Would be a bit cleaner if this part of the code was in its own function

carlosgalvezp commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -79,25 +94,53 @@ def publish(self):
             self.waypoints = lane.waypoints
+            self.n_waypoints = len(lane.waypoints)
+            # Unsubscribe so that waypoint loader stops publishing
+            self.base_waypoints_sub.unregister()
```
I believe I removed this `unregister` some time ago, since Udacity found a better approach

idanov replied:
What is Udacity's approach?

carlosgalvezp replied:
In this commit we merged their changes:
https://github.com/CarND-EuroBots/CarND-Capstone/commit/911e5f040f8f939ef1cc3246069c18ba011a5553

carlosgalvezp commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -164,12 +206,38 @@ def euclidean(cls, pos1, pos2):
+                t = max(d / (prev_vel + MAX_DECEL), .1)
+                old_vel = self.get_waypoint_velocity(wp)
+                vel = prev_vel + MAX_DECEL * t
+                vel = vel if vel > .2 else .0
```
Maybe create a variable for this 0.2? A bit easier to read and keep track of parameters.

idanov replied:
This one is more like a constant which need not be controlled. I will create a constant similar to `MAX_DECEL` (also the latter one will be removed in favour of reading a parameter).

idanov committed:
Address PR comments

 - get max acceleration from the paramter server
 - factor out publishing to rviz in separate methods
 - factor out min velocity to a constant
 - don't unsubscribe from /base_waypoints

carlosgalvezp commented:
Nice, looks really good! I will test it tomorrow :)
